pkg/controller/proxyconfig/validation.go

```diff
@@ -0,0 +1,336 @@
+			_, _, errCIDR := net.ParseCIDR(v)
+			if errDomain != nil && errCIDR != nil {
+				return fmt.Errorf("invalid noProxy: %v", v)
+			}
+		}
+	}
+
+	if isSpecReadinessEndpoints(proxyConfig) {
+		for _, endpoint := range proxyConfig.ReadinessEndpoints {
+			scheme, err := validateURI(endpoint)
+			if err != nil {
+				return fmt.Errorf("invalid URI for endpoint %s: %v", endpoint, err)
+			}
+			switch {
+			case scheme == proxyHTTPScheme:
+				if err := validateHTTPReadinessEndpoint(endpoint); err != nil {
+					return fmt.Errorf("readinessEndpoint probe failed for endpoint %s", endpoint)
+				}
+			case scheme == proxyHTTPSScheme:
+				if !isSpecTrustedCASet(proxyConfig) {
+					return fmt.Errorf("readinessEndpoint with an %q scheme requires trustedCA to be set", proxyHTTPSScheme)
+				}
+				if err := validateHTTPSReadinessEndpoint(readinessCerts, endpoint); err != nil {
+					return fmt.Errorf("readinessEndpoint probe failed for endpoint %s", endpoint)
+				}
+			default:
+				return fmt.Errorf("readiness endpoints requires a %q or %q URI sheme", proxyHTTPScheme, proxyHTTPSScheme)
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateURI validates uri as being a valid url and returns the url scheme.
+func validateURI(uri string) (string, error) {
+	parsed, err := url.Parse(uri)
+	if err != nil {
+		return "", err
+	}
+	if !parsed.IsAbs() {
+		return "", fmt.Errorf("failed validating URI, no scheme for URI %q", uri)
+	}
+	host := parsed.Hostname()
+	if err := validateHost(host); err != nil {
+		return "", fmt.Errorf("failed validating URI %q: %v", uri, err)
+	}
+	if port := parsed.Port(); len(port) != 0 {
+		intPort, err := strconv.Atoi(port)
+		if err != nil {
+			return "", fmt.Errorf("failed converting port to integer for URI %q: %v", uri, err)
+		}
+		if err := validatePort(intPort); err != nil {
+			return "", fmt.Errorf("failed to validate port for URL %q: %v", uri, err)
+		}
+	}
+
+	return parsed.Scheme, nil
+}
+
+// validateHost validates if host is a valid IP address or subdomain in DNS (RFC 1123).
+func validateHost(host string) error {
+	errDomain := validateDomainName(host, false)
+	errIP := validation.IsValidIP(host)
+	if errDomain != nil && errIP != nil {
+		return fmt.Errorf("invalid host: %s", host)
+	}
+
+	return nil
+}
+
+// validatePort validates if port is a valid port number between 1-65535.
+func validatePort(port int) error {
+	invalidPorts := validation.IsValidPortNum(port)
+	if invalidPorts != nil {
+		return fmt.Errorf("invalid port number: %d", port)
+	}
+
+	return nil
+}
+
+// validateHTTPReadinessEndpoint validates an http readinessEndpoint endpoint.
+func validateHTTPReadinessEndpoint(endpoint string) error {
+	if err := validateHTTPReadinessEndpointWithRetries(endpoint, proxyProbeMaxRetries); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateHTTPReadinessEndpointWithRetries tries to validate an http
+// endpoint in a finite loop and returns the last result if it never succeeds.
```

Comment:
I would say that it returns immediately if it fails, no? If it finds a `404`, for example, it returns immediately. Maybe I am not understanding how this is supposed to work.

Reply:
`retries` allows the probe to be run multiple times in case of transient failures. This is a similar approach to pod readiness checks.

Reply:
Hello @danehans . I would say that this function will return on the first failure, even if transient. This function retries only in case of successes.

Reply:
I would say that this function will return on the first failure, even if transient. This function retries only in case of successes.

Reply:
I see the bug in the logic now, thanks for catching this. I'll push an update.

Reply:
It seems like you reverted this one together when you reverted the `return len(proxyConfig.HTTPProxy) > 0` thingie :-)